src/main/java/com/google/devtools/build/lib/rules/cpp/CcCompilationHelper.java

```diff
@@ -1497,12 +1497,19 @@ private Artifact createCompileActionTemplate(
           CppHelper.getDiagnosticsOutputTreeArtifact(
               actionConstructionContext, label, sourceArtifact, outputName, usePic);
     }
+    SpecialArtifact indexstoreTreeArtifact = null;
```

Comment:
I would be against having another `SpecialArtifact` introduced for cc related stuff. Because this is basically un-Starlarkifiable. @oquenchil WDYT?

Reply:
I agree that this should not go in. We shouldn't have any more special casing in native. Someone would need to contribute to move https://github.com/bazelbuild/bazel/issues/17237 forward, then this PR should be based on that or some other generic mechanism that doesn't introduce more native code.

No more SpecialArtifacts should be accepted. https://github.com/bazelbuild/bazel/issues/17237 and Starlark tree artifacts should provide the necessary APIs.

Reply:
Do you perhaps know if aspect with shadow actions can support a use case like this? What I mean is, can you copy CcCompileAction in an aspect and add an output to it, that you can collect?

Reply:
That would mean compiling twice, right? (since we can't have our action be depended on by the original compilation?), where just adding the flag and output only causes a small increase in original compilation time.

Reply:
If you use outputs from both actions, then yes, you compile/cache twice. If you need the outputs for IDE only, then you’d compile once. But the regular compile the users probably do/need couldn’t reuse that.

Replacing outputs in CcInfo might be possible within an extended rule if shadow actions were supported there.

I’m a bit conflicted with the idea that you need to have a complete set of new rules to support IDEs. Or even special toolchains.

If only you could split compilation actions away from parsing+index generation for IDEs that would be ideal solution. That is because I’m assuming parsing is an order faster and because those actions can then run in parallel. You redo some work, but the data for IDEs is ready much sooner.

Reply:
This is the clang/apple way, indexing-while-building. And we use the normal build graph to support that. Adding more actions just for indexing defeats what we are trying to do with this PR.

Reply:
Is there any information on the files that are generated?

Reading clangs docs, it seems the feature is designed to point to a cache directory, where clang also reads the files. 

Is there any concern, that Bazel will use a new empty directory for each file it compiles? This is how tree artifacts work.

This might mean that it will generate an index file for each header file times source file.

What am I missing?

Reply:
Not reflected in this PR yet, but @yongjincho92 took the rules_swift implementation of a "[global index store](https://github.com/bazelbuild/rules_swift/blob/2978c233559361cee087884d13f5c5945c8e6e6f/tools/worker/swift_runner.cc#L144-L181)" and implemented it for this. That makes it (more) efficient. We also plan on changing it to produce an tar file instead of a tree artifact, similar to the comment from @DavidGoldman: https://github.com/bazelbuild/bazel/issues/15983#issuecomment-1272143337 (because that helps with some remote cache performance).

Reply:
"global index store" link seems like it's using a worker. Using a worker makes much more sense, because if I'm not mistaken, you can use a global cache directory for the entire build, that behaves just like expected by clang. It can even read from it. And they are more lax on the hermeticity/sandboxing constraints.

This PR/implementation is touching C++ action that don't have a worker. Bazel has much higher expectations for actions like this. They are executed in a sandbox. I don't see a relation to a worker with this PR. If you implemented a worker, you wouldn't need to pass anything additional parameters to it. Just set an env variable to the path to that directory, or in the toolchain configuration and parse it from IDEs.

Maybe some details are missing? I don't know if it's currently even possible to insert a worker into C++ toolchain.

Reply:
I have a [pr in apple_support](https://github.com/bazelbuild/apple_support/pull/251) that does adopt how rules_swift did [global index store](https://github.com/bazelbuild/rules_swift/commit/914eff948be54cbcde847ae800a2e373e61a5c56).

Reply:
With the change above, I still observed a huge spike in remote cache fetching, so I implemented another change that tars the generated .indexstore directory [here](https://github.com/bazelbuild/apple_support/pull/256/commits/510241167560ed096342b5fda67b47098c805ab9)

Reply:
So after those changes, we just need the ability to declare new outputs, though ideally new arguments as well. The outputs need to be declared in order to be cached properly.